### Deep Learning
#### Visual Product Recommender
#### Learning very non-linear features with neural networks
* It is a classifier, but using more complex, non-linear features.
* Normal classification model is presented below
<img src="Classification_model.png">
* In image classification -> (input features / image pixel -> output -> predicted object)
<img src="Image-classification.png">

* Features are the representation of the data that is used to be fed into the classifier, example -> Text, bag of words, TF-IDF, etc;
* With images there are other representations.

### Neural Networks -> It povides non-linear representation of data (Learning "very" non-linear features);
#### Recap:
* Linear Classifiers -> It creates a line or linear decision boundary between positive class and negative class.
<img src="Linear_classifier.png">
* The score for the linear classifier:
<img src="Linear_classifier_score.png">
* X's -> features, W's -> weights on the features;

#### In neural netwroks, classifiers are represented by graphs, here there is a node for each feature x1, x2, x3, x4,... and a node for the output y(predict).
* Example of the one layer neural network model is presented below:
<img src="Neural_Network_classification_model.png">

#### If the preceptron takes input of exactly 0, what should it output? 
#### Implementaion-defined (An imput of zero is an edge case, there is no hard and fast rule as to whether the preceptron should output 0 or 1. Each implementation can p[ick one wat and outpt the same value for all input values.)

#### What can a linear classifier represent
* Score calculation is presented below
<img src="Linear_classifier_score.png">

* Linear classifier operations:
<img src="Linear_Classifier_Operation.png">

#### A one layer neural network is similar to the linear classifier as both can perform "OR" and "AND" operation. But the simple linear classifier cannot represent "XOR" operations, or they cannot segrate data when there is no line that separates the + and -;
<img src="XOR_classification.png">

#### Here the linear functions will not be useful and therefore non-linear features are useful (neural networks).
<img src="2_layer_neural_netwrok.png">
#### 2 layer neural network;

### Applications of Deep Learning to Computer Vision
* Image features = local detectors, combined to make predictions;
* Local detectors look for "interesting points" in the image;
* Image features - collections of locally interesting points combined to build classifiers;
* Many hand created features exist for finding interest points...
    1. Spin Images
    2. Textons
    3. RIFT
    4. GLOH
    5. HoG
    6. SIFT

#### Standard image classification approach
* Using the hand created image feature - SIFT  feature;
* Input   -> Extract features (create a vector - based on the locations where the firing occured the value exists in the vector)  -> feed it to a simple classifier (e.g., logistioc regression, SVMs) [detect]; 

#### Challenge - Many hand created features exist for finding interest points - but it is a painful process to design.

### Deeplearning - Neural Networks - implicitly learn features
<img src="neural_networks_implict _learning.png">
<img src="Image_Extraction.png">
* Neural networks capture different types of image features at different layers and they get learned automatically.

#### Deep Learning Performance
* Deep learning on ImageNet

### Challenges of Deep Learning
#### Pros
* Enables learning of features rather than hand tuning;
* Impressive performance gains for - Computer Vision, Speech recognition, some text analysis;

#### Challenges
* Requires a lot of data, and labelled data;
* Split into training / validation set -> train on deep neural network -> takes time;
* Computationally expensive, extremely hard to tune;
* Extremely hard to tune
    * Choice of architecture
    * Parameter types
    * Hyperparameters
    * Learning algorithms,
       etc...    
       
### Deep Features
* Deep Features = Deep learning + Transfer learning.
* It allows to build neural networks even in the absence of large amount of data.

#### Transfer learning - use data from one task to help learn on another
<img src="Transfer_learning.png">
* Here the features learnt from cat versus dog get transfered to provide accuracy on the new task, which is detecting 101 categories.

* Since there are many layers the last few are relevant to the specific taks, the last few layers can be chipped of and the rest can then be feed into other algorithms for attaining the predictions.
<img src="Learned_neural_network-1.png">
<img src="Learned_neural_network-2.PNG">
<img src="Learned_neural_network-3.PNG">

### Transfer learning with deep features workflow
<img src="Transfer_learning_deep_features_workflow.png">

### Deep learning ML block diagram
<img src="Deep_Learning_ML_Block_Diagram.png">

### Image Classification with Deep Features

In [1]:
import graphlab

#### Load common image analysis dataset

In [2]:
# Since the dataset is already segregated into train and test set; 
image_train = graphlab.SFrame('image_train_data/')
image_test = graphlab.SFrame('image_test_data/')

This non-commercial license of GraphLab Create for academic use is assigned to amitha353@gmail.com and will expire on May 07, 2019.


[INFO] graphlab.cython.cy_server: GraphLab Create v2.1 started. Logging: C:\Users\Amitha\AppData\Local\Temp\graphlab_server_1526795514.log.0


#### Exploring the image data

In [3]:
graphlab.canvas.set_target('ipynb')

In [5]:
image_train.head()

Columns:
	id	int
	image	Image
	label	str
	deep_features	array
	image_array	array

Rows: 10

Data:
+-----+----------------------+------------+-------------------------------+
|  id |        image         |   label    |         deep_features         |
+-----+----------------------+------------+-------------------------------+
|  24 | Height: 32 Width: 32 |    bird    | [0.242871761322, 1.0954537... |
|  33 | Height: 32 Width: 32 |    cat     | [0.525087952614, 0.0, 0.0,... |
|  36 | Height: 32 Width: 32 |    cat     | [0.566015958786, 0.0, 0.0,... |
|  70 | Height: 32 Width: 32 |    dog     | [1.12979578972, 0.0, 0.0, ... |
|  90 | Height: 32 Width: 32 |    bird    | [1.71786928177, 0.0, 0.0, ... |
|  97 | Height: 32 Width: 32 | automobile | [1.57818555832, 0.0, 0.0, ... |
| 107 | Height: 32 Width: 32 |    dog     | [0.0, 0.0, 0.220677852631,... |
| 121 | Height: 32 Width: 32 |    bird    | [0.0, 0.23753464222, 0.0, ... |
| 136 | Height: 32 Width: 32 | automobile | [0.0, 0.0, 0.0, 0.0, 0.0, ... |
| 138 | Height: 32 Width: 32 |    bird    | [0.658935725689, 0.0, 0.0,... |
+-----+----------------------+------------+-------------------------------+
+-------------------------------+
|          image_array          |
+-------------------------------+
| [73.0, 77.0, 58.0, 71.0, 6... |
| [7.0, 5.0, 8.0, 7.0, 5.0, ... |
| [169.0, 122.0, 65.0, 131.0... |
| [154.0, 179.0, 152.0, 159.... |
| [216.0, 195.0, 180.0, 201.... |
| [33.0, 44.0, 27.0, 29.0, 4... |
| [97.0, 51.0, 31.0, 104.0, ... |
| [93.0, 96.0, 88.0, 102.0, ... |
| [35.0, 59.0, 53.0, 36.0, 5... |
| [205.0, 193.0, 195.0, 200.... |
+-------------------------------+
[10 rows x 5 columns]

In [6]:
image_train['image'].show()

#### Train classifier on the raw image pixels (Features -> raw pixels and not deep features) 

In [7]:
raw_pixel_model = graphlab.logistic_classifier.create(image_train, 
                                               target='label', 
                                               features=['image_array'])

PROGRESS: Creating a validation set from 5 percent of training data. This may take a while.
          You can set ``validation_set=None`` to disable validation tracking.



WARNING: The number of feature dimensions in this problem is very large in comparison with the number of examples. Unless an appropriate regularization value is set, this model may not provide accurate predictions for a validation/test set.

Logistic regression:

--------------------------------------------------------

Number of examples          : 1904

Number of classes           : 4

Number of feature columns   : 1

Number of unpacked features : 3072

Number of coefficients    : 9219

Starting L-BFGS

--------------------------------------------------------

+-----------+----------+-----------+--------------+-------------------+---------------------+

| Iteration | Passes   | Step size | Elapsed Time | Training-accuracy | Validation-accuracy |

+-----------+----------+-----------+--------------+-------------------+---------------------+

| 1         | 6        | 0.000026  | 3.708570     | 0.351366          | 0.316832            |

| 2         | 8        | 1.000000  | 4.599208     | 0.418067          | 0.326733            |

| 3         | 9        | 1.000000  | 5.021090     | 0.440651          | 0.346535            |

| 4         | 10       | 1.000000  | 5.396096     | 0.447479          | 0.366337            |

| 5         | 11       | 1.000000  | 5.786725     | 0.445903          | 0.326733            |

| 6         | 12       | 1.000000  | 6.146105     | 0.467437          | 0.376238            |

| 10        | 16       | 1.000000  | 7.302374     | 0.540966          | 0.425743            |

+-----------+----------+-----------+--------------+-------------------+---------------------+

TERMINATED: Iteration limit reached.

This model may not be optimal. To improve it, consider increasing `max_iterations`.

#### Make prediction with the simple model based on raw pixels

In [8]:
image_test[0:3]['image'].show()

In [9]:
image_test[0:3]['label']

dtype: str
Rows: 3
['cat', 'automobile', 'cat']

In [10]:
raw_pixel_model.predict(image_test[0:3])

dtype: str
Rows: 3
['bird', 'cat', 'bird']

#### Evaluating raw pixel model on test-data

In [11]:
### Evaluate - computes error metrics.
raw_pixel_model.evaluate(image_test)

{'accuracy': 0.484, 'auc': 0.7299180416666669, 'confusion_matrix': Columns:
 	target_label	str
 	predicted_label	str
 	count	int
 
 Rows: 16
 
 Data:
 +--------------+-----------------+-------+
 | target_label | predicted_label | count |
 +--------------+-----------------+-------+
 |     bird     |       dog       |  198  |
 |     dog      |       cat       |  195  |
 |     bird     |       cat       |  142  |
 |     dog      |       dog       |  460  |
 |     cat      |       dog       |  352  |
 |     dog      |    automobile   |   84  |
 |     bird     |    automobile   |  114  |
 |  automobile  |    automobile   |  612  |
 |     cat      |       bird      |  202  |
 |     cat      |    automobile   |  128  |
 +--------------+-----------------+-------+
 [16 rows x 3 columns]
 Note: Only the head of the SFrame is printed.
 You can use print_rows(num_rows=m, num_columns=n) to print more rows and columns., 'f1_score': 0.4823025015419202, 'log_loss': 1.1909685428530576, 'precision': 0.4

#### The accuracy of the above model is veru low - 48%

#### Improving the model using the Deep Features (Using Transfer Learning)

In [12]:
len(image_train)

2005

In [14]:
# Below 2 commands perform transfer learning - take the imagenet training on neural network and apply to the new (Safari 10) dataset.
# deep_learning_model = graphlab.load_model('imagenet_model')
# image_train['deep_features'] = deep_learning_model.extract_features(image_train)

In [15]:
image_train.head()

Columns:
	id	int
	image	Image
	label	str
	deep_features	array
	image_array	array

Rows: 10

Data:
+-----+----------------------+------------+-------------------------------+
|  id |        image         |   label    |         deep_features         |
+-----+----------------------+------------+-------------------------------+
|  24 | Height: 32 Width: 32 |    bird    | [0.242871761322, 1.0954537... |
|  33 | Height: 32 Width: 32 |    cat     | [0.525087952614, 0.0, 0.0,... |
|  36 | Height: 32 Width: 32 |    cat     | [0.566015958786, 0.0, 0.0,... |
|  70 | Height: 32 Width: 32 |    dog     | [1.12979578972, 0.0, 0.0, ... |
|  90 | Height: 32 Width: 32 |    bird    | [1.71786928177, 0.0, 0.0, ... |
|  97 | Height: 32 Width: 32 | automobile | [1.57818555832, 0.0, 0.0, ... |
| 107 | Height: 32 Width: 32 |    dog     | [0.0, 0.0, 0.220677852631,... |
| 121 | Height: 32 Width: 32 |    bird    | [0.0, 0.23753464222, 0.0, ... |
| 136 | Height: 32 Width: 32 | automobile | [0.0, 0.0, 0.0, 0.0, 0.0, ... |
| 138 | Height: 32 Width: 32 |    bird    | [0.658935725689, 0.0, 0.0,... |
+-----+----------------------+------------+-------------------------------+
+-------------------------------+
|          image_array          |
+-------------------------------+
| [73.0, 77.0, 58.0, 71.0, 6... |
| [7.0, 5.0, 8.0, 7.0, 5.0, ... |
| [169.0, 122.0, 65.0, 131.0... |
| [154.0, 179.0, 152.0, 159.... |
| [216.0, 195.0, 180.0, 201.... |
| [33.0, 44.0, 27.0, 29.0, 4... |
| [97.0, 51.0, 31.0, 104.0, ... |
| [93.0, 96.0, 88.0, 102.0, ... |
| [35.0, 59.0, 53.0, 36.0, 5... |
| [205.0, 193.0, 195.0, 200.... |
+-------------------------------+
[10 rows x 5 columns]

#### Given the deep features, lets train a classifier

In [16]:
deep_feaures_model = graphlab.logistic_classifier.create(image_train,
                                                        features=['deep_features'],
                                                        target='label')

PROGRESS: Creating a validation set from 5 percent of training data. This may take a while.
          You can set ``validation_set=None`` to disable validation tracking.



WARNING: The number of feature dimensions in this problem is very large in comparison with the number of examples. Unless an appropriate regularization value is set, this model may not provide accurate predictions for a validation/test set.

WARNING: Detected extremely low variance for feature(s) 'deep_features' because all entries are nearly the same.
Proceeding with model training using all features. If the model does not provide results of adequate quality, exclude the above mentioned feature(s) from the input dataset.

Logistic regression:

--------------------------------------------------------

Number of examples          : 1918

Number of classes           : 4

Number of feature columns   : 1

Number of unpacked features : 4096

Number of coefficients    : 12291

Starting L-BFGS

--------------------------------------------------------

+-----------+----------+-----------+--------------+-------------------+---------------------+

| Iteration | Passes   | Step size | Elapsed Time | Training-accuracy | Validation-accuracy |

+-----------+----------+-----------+--------------+-------------------+---------------------+

| 1         | 5        | 0.000130  | 2.937540     | 0.748697          | 0.701149            |

| 2         | 9        | 0.250000  | 4.562565     | 0.773723          | 0.655172            |

| 3         | 10       | 0.250000  | 4.984444     | 0.779458          | 0.655172            |

| 4         | 11       | 0.250000  | 5.406324     | 0.784672          | 0.666667            |

| 5         | 12       | 0.250000  | 5.796956     | 0.786757          | 0.678161            |

| 6         | 13       | 0.250000  | 6.234462     | 0.802920          | 0.689655            |

| 10        | 17       | 0.250000  | 7.796985     | 0.887383          | 0.712644            |

+-----------+----------+-----------+--------------+-------------------+---------------------+

TERMINATED: Iteration limit reached.

This model may not be optimal. To improve it, consider increasing `max_iterations`.

#### Apply the deep features model to the first few images in the test set

In [17]:
image_test[0:3]['image'].show()

In [18]:
deep_feaures_model.predict(image_test[0:3])

dtype: str
Rows: 3
['cat', 'automobile', 'cat']

#### Compute the test data accuracy of the deep_feature_model

In [19]:
deep_feaures_model.evaluate(image_test)

{'accuracy': 0.79025, 'auc': 0.9426577083333353, 'confusion_matrix': Columns:
 	target_label	str
 	predicted_label	str
 	count	int
 
 Rows: 16
 
 Data:
 +--------------+-----------------+-------+
 | target_label | predicted_label | count |
 +--------------+-----------------+-------+
 |     dog      |    automobile   |   14  |
 |     bird     |       bird      |  791  |
 |  automobile  |       bird      |   18  |
 |     dog      |       bird      |   38  |
 |     cat      |    automobile   |   35  |
 |     cat      |       cat       |  682  |
 |     bird     |       cat       |  126  |
 |     cat      |       dog       |  219  |
 |     dog      |       dog       |  734  |
 |  automobile  |    automobile   |  954  |
 +--------------+-----------------+-------+
 [16 rows x 3 columns]
 Note: Only the head of the SFrame is printed.
 You can use print_rows(num_rows=m, num_columns=n) to print more rows and columns., 'f1_score': 0.7909674044256507, 'log_loss': 0.5510528442111333, 'precision': 0

#### Accuracy of the Deep features model is better - Accuracy = 79%